### ISSUE TITLE
Transfer connection initiation following an unallowed (queued) transfer response

### ISSUE THREAD

**Issue by jpdillingham** [Nicotine-Plus/nicotine-plus#653]
In short, I'm unable to download from Nicotine+ with Soulseek.NET if the initial request for the transfer was disallowed.  I initially thought this may be a connection issue based on Nicotine+ logs (there's a message about being unable to write to a closed connection, but the timing is very weird), but I've been able to confirm that transfers which are allowed and can thus begin immediately are working fine.

The impacted lines in my code are [here](https://github.com/jpdillingham/Soulseek.NET/blob/master/src/Soulseek/SoulseekClient.cs#L1988-L2047).  The block following `transferRequestAcknowledgement.IsAllowed` is working, and the `else` is not.

I was able to reproduce the problem consistently by configuring Nicotine+ with a queue limit of 0, requesting a file, then increasing the limit to 1.  Here's an excerpt of my debug logs while I attempted this, starting with the messages produced when the file was originally requested:

```
// negotiating a message connection
[09:44:37.729] [DIAGNOSTIC:ServerMessageHandler] [Debug] Server message received: GetPeerAddress
[09:44:37.733] [DIAGNOSTIC:SoulseekClient] [Debug] EndPoint cache MISS for nicotine+: 173.30.82.88:2234
[09:44:37.738] [DIAGNOSTIC:PeerConnectionManager] [Debug] Attempting simultaneous direct and indirect message connections to nicotine+ (173.30.82.88:2234)
[09:44:37.740] [DIAGNOSTIC:PeerConnectionManager] [Debug] Attempting direct message connection to nicotine+ (173.30.82.88:2234)
[09:44:37.744] [DIAGNOSTIC:PeerConnectionManager] [Debug] Soliciting indirect message connection to nicotine+ with token 1
[09:44:37.745] [DIAGNOSTIC:PeerConnectionManager] [Debug] Direct message connection to nicotine+ (173.30.82.88:2234) established. (type: Outbound, Direct, id: 7289d258-7be0-45fe-b16b-b6d1aa1c21ec)
[09:44:37.749] [DIAGNOSTIC:PeerConnectionManager] [Debug] Direct message connection to nicotine+ (173.30.82.88:2234) established first, attempting to cancel indirect connection.
[09:44:37.761] [DIAGNOSTIC:PeerConnectionManager] [Debug] Message connection to nicotine+ (173.30.82.88:2234) established. (type: Outbound, Direct, id: 7289d258-7be0-45fe-b16b-b6d1aa1c21ec)
[09:44:37.762] [DIAGNOSTIC:PeerConnectionManager] [Debug] Failed to establish an indirect message connection to nicotine+ with token 1: A task was canceled.
// event fired immediately after writing the request to the message connection
[DOWNLOAD] [nicotine+/album.nfo] None => Requested
// unsolicited connection attempted from Nicotine+ fails because it is unexpected until the transfer response is allowed
[09:44:37.812] [DIAGNOSTIC:ListenerHandler] [Debug] Accepted incoming connection from 192.168.1.1:54859 (id: f8878af1-84b5-47e8-a2b8-99c1bb0a10bf)
[09:44:38.317] [DIAGNOSTIC:ListenerHandler] [Debug] Failed to initialize direct connection from 192.168.1.1:54354: Unknown PierceFirewall attempt with token 1 from 192.168.1.1:54354 (id: f8878af1-84b5-47e8-a2b8-99c1bb0a10bf)
// transfer response from Nicotine+ with disallowed status
[09:44:38.317] [DIAGNOSTIC:PeerMessageHandler] [Debug] Peer message received: TransferResponse from nicotine+ (173.30.82.88:2234) (id: 7289d258-7be0-45fe-b16b-b6d1aa1c21ec)
// event fired immediately after transfer is disallowed and therefore queued locally
[DOWNLOAD] [nicotine+/album.nfo] Requested => Queued
[09:44:53.323] [DIAGNOSTIC:PeerConnectionManager] [Debug] Message connection to nicotine+ (173.30.82.88:2234) disconnected. (type: Outbound, Direct, id: 7289d258-7be0-45fe-b16b-b6d1aa1c21ec)
[09:44:53.325] [DIAGNOSTIC:PeerConnectionManager] [Debug] Removed message connection record for nicotine+ (173.30.82.88:2234) (type: Outbound, Direct, id: 7289d258-7be0-45fe-b16b-b6d1aa1c21ec)
[09:44:53.326] [DIAGNOSTIC:PeerConnectionManager] [Debug] Message connection cache now contains 0 connections.
```

Here's the log after the Nicotine+ queue limit is raised to 1:

```
// negotiating a message connection
[09:45:10.361] [DIAGNOSTIC:ServerMessageHandler] [Debug] Server message received: ConnectToPeer
[09:45:10.362] [DIAGNOSTIC:ServerMessageHandler] [Debug] Received message ConnectToPeer request from nicotine+ (173.30.82.88:2234)
[09:45:10.366] [DIAGNOSTIC:PeerConnectionManager] [Debug] Attempting inbound indirect message connection to nicotine+ (173.30.82.88:2234) for token 269
[09:45:10.370] [DIAGNOSTIC:PeerConnectionManager] [Debug] Message connection to nicotine+ (173.30.82.88:2234) established. (type: Inbound, Indirect, id: c085ae27-743e-4c70-a474-12d59f1e468b)
// transfer requested
[09:45:10.873] [DIAGNOSTIC:PeerMessageHandler] [Debug] Peer message received: TransferRequest from nicotine+ (173.30.82.88:2234) (id: c085ae27-743e-4c70-a474-12d59f1e468b)
// event fired after transfer is requested
[DOWNLOAD] [nicotine+/album.nfo] Queued => Initializing
// fetching the existing message connection to send the response
[09:45:10.879] [DIAGNOSTIC:PeerConnectionManager] [Debug] Retrieved cached message connection to nicotine+ (173.30.82.88:2234) (type: Inbound, Indirect, id: c085ae27-743e-4c70-a474-12d59f1e468b)
// expecting Nicotine+ to initiate a transfer connection to send the file
[09:45:10.881] [DIAGNOSTIC:PeerConnectionManager] [Debug] Waiting for a direct or indirect transfer connection from nicotine+ with remote token 268 for shared\killing your enemy in 1995\album.nfo
[09:45:15.902] [DIAGNOSTIC:PeerConnectionManager] [Debug] Failed to establish a direct or indirect transfer connection to nicotine+ with remote token 268 for shared\killing your enemy in 1995\album.nfo
[09:45:15.932] [DIAGNOSTIC:SoulseekClient] [Debug] Soulseek.Exceptions.ConnectionException: Failed to establish a direct or indirect transfer connection to nicotine+ with remote token 268 for shared\killing your enemy in 1995\album.nfo
   at Soulseek.Network.PeerConnectionManager.AwaitTransferConnectionAsync(String username, String filename, Int32 remoteToken, CancellationToken cancellationToken) in C:\Users\JP.WHATNET\source\repos\Soulseek.NET\src\Soulseek\Network\PeerConnectionManager.cs:line 247
   at Soulseek.SoulseekClient.DownloadToStreamAsync(String username, String filename, Stream outputStream, Nullable`1 size, Int64 startOffset, Int32 token, TransferOptions options, CancellationToken cancellationToken) in C:\Users\JP.WHATNET\source\repos\Soulseek.NET\src\Soulseek\SoulseekClient.cs:line 2049
[DOWNLOAD] [nicotine+/album.nfo] Initializing => Completed, Errored (0/2249 = 0%)
```

Judging by the logs above only, it looks as though the outbound transfer connection to upload the file is established when the file is queued, not when the request/response handshake is completed.

It's been a while since I wrote this code but I seem to recall that there's a few ways to negotiate a transfer with the official client; requesting that a file be queued up front, or just requesting the transfer and inspecting the response to see whether it was queued.  I opted for the latter and I'm wondering if Nicotine+ isn't expecting the former.

**Comment by hboetes** [comment]
Can you reproduce this problem with the latest version from git?

**Comment by jpdillingham** [comment]
Yep.  The above logs are from a test with 1.4.x, and I've just now tested with the latest version from the website, and the build artifact from the last commit [here](https://github.com/Nicotine-Plus/nicotine-plus/actions/runs/248778845).  Same results each time.

**Comment by mathiascode** [comment]
I'm a bit rusty when it comes to the transfer code, but these should be the relevant parts:
- Initial response if upload is allowed: https://github.com/Nicotine-Plus/nicotine-plus/blob/master/pynicotine/transfers.py#L617
- Later upload when the queue is empty: https://github.com/Nicotine-Plus/nicotine-plus/blob/master/pynicotine/transfers.py#L1566, which initiates a new transfer request here: https://github.com/Nicotine-Plus/nicotine-plus/blob/66f4767a8840b32132886819750ee638d82f2a46/pynicotine/transfers.py#L303

**Comment by mathiascode** [comment]
Is this still something that needs to be changed on the Nicotine+ side?

**Comment by jpdillingham** [comment]
I found a couple of interesting things but I don't think either are bugs.

Regarding the problem that raised this issue, what I discovered is that Nicotine+ is expecting the remote client to initiate the download, rather than initiating itself, so I added a fallback to initiate the transfer in that case: https://github.com/jpdillingham/Soulseek.NET/blob/master/src/Soulseek/SoulseekClient.cs#L2040-L2058

I tried to duplicate this behavior with both NS and QT and was unable, but it seems like it must be supported otherwise Nicotine+ wouldn't work with them.  I suspect this may be a subtle difference that correlates with queueing transfers first instead of just requesting them.

I found another issue with uploads when waiting for a connection to disconnect, which I solved by adding a "linger": https://github.com/jpdillingham/Soulseek.NET/blob/master/src/Soulseek/SoulseekClient.cs#L2727-L2749.  This has always been a contentious issue for me since the code is async rather than event driven and I need to return when the transfer is done, but closing the connection too early will cause a RST before all of the data is transferred.  With NS and QT I simply try to read 1 byte from the connection and that reliably fails and closes the connection, but Nicotine+ happily returns `0x0` for 400-500ms, presumably while it is still reading its input buffer.  

Thanks for the help!
